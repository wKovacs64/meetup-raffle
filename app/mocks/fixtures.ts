import { faker } from '@faker-js/faker';

//
// API routes
//
export const EVENTS_ENDPOINT = 'https://api.meetup.com/:meetup/events';
export const RSVPS_ENDPOINT =
  'https://api.meetup.com/:meetup/events/:eventId/rsvps';

//
// Mock responses
//
export const UPCOMING_EVENTS = [
  { visibility: 'public', id: faker.random.alphaNumeric(13) },
];

export const EVENT_RSVPS = [
  ...generateOrganizers(1),
  ...generateCoOrganizers(1),
  ...generateMembers(20),
];

//
// Convenience functions to generate data
//
function generateOrganizers(count: number) {
  return Array.from(Array(count), () => ({
    response: 'yes',
    member: {
      id: faker.datatype.number({ min: 100000000, max: 999999999 }),
      name: faker.name.firstName(),
      role: 'organizer',
      event_context: { host: true },
    },
    group: { urlname: faker.company.bsNoun() },
  }));
}

function generateCoOrganizers(count: number) {
  return Array.from(Array(count), () => ({
    response: 'yes',
    member: {
      id: faker.datatype.number({ min: 100000000, max: 999999999 }),
      name: faker.name.fullName(),
      photo: {
        id: faker.datatype.number({ min: 100000000, max: 999999999 }),
        highres_link: faker.internet.avatar(),
        photo_link: faker.internet.avatar(),
        thumb_link: faker.internet.avatar(),
        type: 'member',
        base_url: 'https://secure.meetupstatic.com',
      },
      role: 'coorganizer',
      event_context: { host: true },
    },
    group: { urlname: faker.company.bsNoun() },
  }));
}

function generateMembers(count: number) {
  return Array.from(Array(count), () => ({
    response: 'yes',
    member: {
      id: faker.datatype.number({ min: 100000000, max: 999999999 }),
      name: faker.name.fullName(),
      photo: {
        id: faker.datatype.number({ min: 100000000, max: 999999999 }),
        highres_link: faker.internet.avatar(),
        photo_link: faker.internet.avatar(),
        thumb_link: faker.internet.avatar(),
        type: 'member',
        base_url: 'https://secure.meetupstatic.com',
      },
      type: 'member',
      event_context: { host: false },
    },
    group: { urlname: faker.company.bsNoun() },
  }));
}
