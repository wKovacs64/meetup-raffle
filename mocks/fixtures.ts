import { faker } from '@faker-js/faker';

//
// API routes
//
export const EVENTS_ENDPOINT = 'https://api.meetup.com/:meetup/events';
export const RSVPS_ENDPOINT =
  'https://api.meetup.com/:meetup/events/:eventId/rsvps';

//
// Mock responses
//
export const UPCOMING_EVENTS = [
  { visibility: 'public', id: faker.string.alphanumeric(13) },
];

export const EVENT_RSVPS = [
  ...generateOrganizers(1),
  ...generateCoOrganizers(1),
  ...generateMembers(20),
];

//
// Convenience functions to generate data
//
function generateOrganizers(count: number) {
  return Array.from(Array(count), () => ({
    response: 'yes',
    member: {
      id: faker.number.int({ min: 100000000, max: 999999999 }),
      name: faker.person.firstName(),
      role: 'organizer',
      event_context: { host: true },
    },
    group: { urlname: faker.company.buzzNoun() },
  }));
}

function generateCoOrganizers(count: number) {
  return Array.from(Array(count), () => ({
    response: 'yes',
    member: {
      id: faker.number.int({ min: 100000000, max: 999999999 }),
      name: faker.person.fullName(),
      photo: {
        id: faker.number.int({ min: 100000000, max: 999999999 }),
        highres_link: faker.internet.avatar(),
        photo_link: faker.internet.avatar(),
        thumb_link: faker.internet.avatar(),
        type: 'member',
        base_url: 'https://secure.meetupstatic.com',
      },
      role: 'coorganizer',
      event_context: { host: true },
    },
    group: { urlname: faker.company.buzzNoun() },
  }));
}

function generateMembers(count: number) {
  return Array.from(Array(count), () => ({
    response: 'yes',
    member: {
      id: faker.number.int({ min: 100000000, max: 999999999 }),
      name: faker.person.fullName(),
      photo: {
        id: faker.number.int({ min: 100000000, max: 999999999 }),
        highres_link: faker.internet.avatar(),
        photo_link: faker.internet.avatar(),
        thumb_link: faker.internet.avatar(),
        type: 'member',
        base_url: 'https://secure.meetupstatic.com',
      },
      type: 'member',
      event_context: { host: false },
    },
    group: { urlname: faker.company.buzzNoun() },
  }));
}
