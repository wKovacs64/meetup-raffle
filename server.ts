import * as build from '@remix-run/dev/server-build';
import { createRequestHandler } from '@netlify/remix-adapter';
import { installGlobals } from '@remix-run/node';
import sourceMapSupport from 'source-map-support';

sourceMapSupport.install();
installGlobals();

export const handler = createRequestHandler({
  // @ts-ignore
  build,
  mode: build.mode,
});
