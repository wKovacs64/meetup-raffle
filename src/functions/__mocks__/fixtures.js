export const EVENT_ID = 'foo123';
export const MEETUP = 'foo';
