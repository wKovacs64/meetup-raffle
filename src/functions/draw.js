import axios from 'axios';
import meetupRandomizer from 'meetup-randomizer';
import {
  getParamsFromRequest,
  getIdFromEvent,
  parseEventsResponse,
  validateStatus,
} from './helpers';

const handler = async (request /* , context */) => {
  const headers =
    process.env.NODE_ENV === 'development'
      ? { 'Access-Control-Allow-Origin': '*' }
      : {};

  let meetup;
  let count;
  let specificEventId;

  try {
    ({ meetup, count, specificEventId } = getParamsFromRequest(request));
  } catch (err) {
    return {
      headers,
      statusCode: 400,
      body: JSON.stringify({ error: { message: err.message } }),
    };
  }

  try {
    const baseUrl = `https://api.meetup.com/${encodeURIComponent(
      meetup,
    )}/events`;
    const eventUrlSuffix = '?status=upcoming&only=id,visibility';
    const eventUrl = `${baseUrl}/${encodeURIComponent(
      specificEventId,
    )}${eventUrlSuffix}`;

    const response = await axios.get(eventUrl, { validateStatus });
    const eventId = getIdFromEvent(parseEventsResponse(response));
    const winners = await meetupRandomizer.run(meetup, eventId, count);

    if (Array.isArray(winners) && winners.length) {
      return {
        headers,
        statusCode: 200,
        body: JSON.stringify({ winners }),
      };
    }

    throw new Error('Sorry, we received unexpected data for that request.');
  } catch (err) {
    return {
      headers,
      statusCode: 404,
      body: JSON.stringify({ error: { message: err.message } }),
    };
  }
};

export { handler };
