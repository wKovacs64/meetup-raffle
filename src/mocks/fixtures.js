import faker from 'faker';

// API routes
export const EVENTS_ENDPOINT = 'https://api.meetup.com/:meetup/events';
export const RSVPS_ENDPOINT =
  'https://api.meetup.com/:meetup/events/:eventId/rsvps';

// Initialize some data up front for reusability
const eventId = faker.random.alphaNumeric(13);
const meetupName = 'some-meetup';

// Mock responses
export const UPCOMING_EVENTS = [{ visibility: 'public', id: eventId }];
export const EVENT_RSVPS = [
  {
    response: 'yes',
    member: {
      id: faker.datatype.number(100000000, 999999999),
      name: faker.name.firstName(),
      role: 'organizer',
      event_context: { host: true },
    },
    group: { urlname: meetupName },
  },
  {
    response: 'yes',
    member: {
      id: faker.datatype.number(100000000, 999999999),
      name: faker.name.findName(),
      photo: {
        id: faker.datatype.number(100000000, 999999999),
        highres_link: faker.internet.avatar(),
        photo_link: faker.internet.avatar(),
        thumb_link: faker.internet.avatar(),
        type: 'member',
        base_url: 'https://secure.meetupstatic.com',
      },
      role: 'coorganizer',
      event_context: { host: true },
    },
    group: { urlname: meetupName },
  },
  {
    response: 'yes',
    member: {
      id: faker.datatype.number(100000000, 999999999),
      name: faker.name.findName(),
      photo: {
        id: faker.datatype.number(100000000, 999999999),
        highres_link: faker.internet.avatar(),
        photo_link: faker.internet.avatar(),
        thumb_link: faker.internet.avatar(),
        type: 'member',
        base_url: 'https://secure.meetupstatic.com',
      },
      type: 'member',
      event_context: { host: false },
    },
    group: { urlname: meetupName },
  },
  {
    response: 'yes',
    member: {
      id: faker.datatype.number(100000000, 999999999),
      name: faker.name.findName(),
      photo: {
        id: faker.datatype.number(100000000, 999999999),
        highres_link: faker.internet.avatar(),
        photo_link: faker.internet.avatar(),
        thumb_link: faker.internet.avatar(),
        type: 'member',
        base_url: 'https://secure.meetupstatic.com',
      },
      type: 'member',
      event_context: { host: false },
    },
    group: { urlname: meetupName },
  },
];
