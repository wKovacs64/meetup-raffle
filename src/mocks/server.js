import { setupServer } from 'msw/node';
import { handlers } from './handlers';

// Setup Node (Jest) request interception using the given mocks.
export const server = setupServer(...handlers);

// some tests import msw exports from this file
export * from 'msw';
