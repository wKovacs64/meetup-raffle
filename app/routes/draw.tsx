import { json, type LoaderArgs } from '@remix-run/node';
import { Link, useLoaderData, useNavigation } from '@remix-run/react';
import { z, ZodError } from '~/vendor/zod.server';
import { meetupRandomizer } from '~/vendor/meetup-randomizer.server';
import { userSettingsCookie } from '~/core/cookies.server';
import { getEventFromResponseData } from '~/raffle/get-event-from-response-data';
import { getIdFromEvent } from '~/raffle/get-id-from-event';
import LoadingSpinner from '~/raffle/loading-spinner';
import Winners from '~/raffle/winners';
import ErrorMessage from '~/raffle/error-message';
import type { UpcomingEvent, Winner } from '~/types';

export async function loader({ request }: LoaderArgs) {
  const formSchema = z.object({
    count: z.string().refine(
      (value) => {
        const valueAsNumber = Number.parseInt(value, 10);
        const isNumber = !Number.isNaN(valueAsNumber);
        const isInRange = valueAsNumber > 0 && valueAsNumber < 10;
        return isNumber && isInRange;
      },
      {
        message: 'Count must be a number between 1 and 9',
      },
    ),
    meetup: z.string().min(1, 'Meetup name is required'),
  });

  let userSettings = request.headers.get('Cookie') ?? '';

  const { searchParams } = new URL(request.url);
  const formData = {
    meetup: searchParams.get('meetup') ?? '',
    count: searchParams.get('count') ?? '',
  };

  let meetup;
  let count;

  try {
    const validatedFormData = formSchema.parse(formData);
    count = validatedFormData.count;
    meetup = validatedFormData.meetup;
    userSettings = await userSettingsCookie.serialize(
      { meetup, count },
      {
        // 10 years
        expires: new Date(Date.now() + 10 * 365 * 24 * 60 * 60 * 1000),
      },
    );
  } catch (err) {
    if (err instanceof ZodError) console.error(JSON.stringify(err.issues));
    return jsonWithCookie(
      {
        formData,
        errorMessage:
          'Sorry, somehow the form was submitted with invalid data.',
      },
      userSettings,
    );
  }

  const eventUrl = `https://api.meetup.com/${encodeURIComponent(
    meetup,
  )}/events?status=upcoming&only=id,visibility`;

  const res = await fetch(eventUrl);

  // This case covers invalid Meetup group names as well as invalid event IDs.
  if (res.status === 404) {
    return jsonWithCookie(
      {
        formData,
        errorMessage: "Sorry, I couldn't find any information on that.",
      },
      userSettings,
    );
  }

  const data: UpcomingEvent | Array<UpcomingEvent> = await res.json();
  const event = getEventFromResponseData(data);

  if (!event) {
    return jsonWithCookie(
      {
        formData,
        errorMessage: "Sorry, I couldn't find any upcoming events.",
      },
      userSettings,
    );
  }

  const eventId = getIdFromEvent(event);

  if (eventId === null) {
    return jsonWithCookie(
      {
        formData,
        errorMessage: 'Sorry, their members list is private.',
      },
      userSettings,
    );
  }

  try {
    const winners: Array<Winner> = await meetupRandomizer.run(
      meetup,
      eventId,
      count,
    );

    if (Array.isArray(winners) && winners.length) {
      return jsonWithCookie(
        {
          formData,
          winners,
        },
        userSettings,
      );
    }
  } catch (err) {
    const isError = err instanceof Error;
    return jsonWithCookie(
      {
        formData,
        errorMessage: isError ? err.message : 'Sorry, something went wrong.',
      },
      userSettings,
    );
  }

  return jsonWithCookie(
    {
      formData,
      errorMessage: 'Sorry, we received unexpected data for that request.',
    },
    userSettings,
  );
}

function jsonWithCookie<T>(
  data: Parameters<typeof json<T>>[0],
  cookieString: string,
) {
  return json<T>(data, {
    headers: {
      'Set-Cookie': cookieString,
    },
  });
}

export default function DrawPage({ onRetry }: { onRetry: () => void }) {
  const data = useLoaderData<typeof loader>();
  const searchParams = new URLSearchParams({
    meetup: data.formData.meetup,
    count: data.formData.count,
  });

  if (useNavigation().state === 'loading') return <LoadingSpinner />;

  return (
    <div className="mt-4 sm:mt-8">
      {'winners' in data ? (
        <Winners winners={data.winners} />
      ) : 'errorMessage' in data ? (
        <ErrorMessage problemText={data.errorMessage} />
      ) : null}
      <div className="my-8 flex flex-shrink-0 flex-wrap justify-around">
        <Link
          to=".."
          className="mb-4 w-full border border-solid border-current bg-white px-16 py-4 text-center font-bold shadow-lg hover:bg-gray-300 focus:bg-gray-300 sm:mb-0 sm:w-64 sm:text-xl"
          replace
        >
          Start Over
        </Link>
        <Link
          to={`.?${searchParams.toString()}`}
          className="w-full border border-solid border-current bg-white px-16 py-4 text-center font-bold shadow-lg hover:bg-gray-300 focus:bg-gray-300 sm:w-64 sm:text-xl"
          replace
        >
          Draw Again
        </Link>
      </div>
    </div>
  );
}
