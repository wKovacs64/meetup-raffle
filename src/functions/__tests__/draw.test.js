import meetupRandomizer from 'meetup-randomizer';
import { server, rest } from '../../test/server';
import { EVENTS_ENDPOINT } from '../../test/fixtures';
import { handler } from '../draw';

const draw = async ({ meetup, count = 1 }) =>
  handler(
    {
      httpMethod: 'GET',
      queryStringParameters: { meetup, count },
    },
    {},
  );

describe('draw', () => {
  it('handles invalid requests', async () => {
    expect(await draw({})).toMatchInlineSnapshot(`
      Object {
        "body": "{\\"error\\":{\\"message\\":\\"The \\\\\\"meetup\\\\\\" query parameter is required and must be a string.\\"}}",
        "headers": Object {},
        "statusCode": 400,
      }
    `);
  });

  it('handles Meetup or Event not found', async () => {
    server.use(
      rest.get(EVENTS_ENDPOINT, (req, res, ctx) => {
        return res.once(ctx.status(404));
      }),
    );

    expect(await draw({ meetup: 'not-found' })).toMatchInlineSnapshot(`
      Object {
        "body": "{\\"error\\":{\\"message\\":\\"Sorry, I couldn't find any information on that.\\"}}",
        "headers": Object {},
        "statusCode": 404,
      }
    `);
  });

  it('handles no upcoming Events found', async () => {
    server.use(
      rest.get(EVENTS_ENDPOINT, (req, res, ctx) => {
        return res.once(ctx.json([]));
      }),
    );

    expect(await draw({ meetup: 'no-events' })).toMatchInlineSnapshot(`
      Object {
        "body": "{\\"error\\":{\\"message\\":\\"Sorry, I couldn't find any upcoming events.\\"}}",
        "headers": Object {},
        "statusCode": 404,
      }
    `);
  });

  it('handles a limited visibility Event', async () => {
    server.use(
      rest.get(EVENTS_ENDPOINT, (req, res, ctx) => {
        return res.once(
          ctx.json([
            {
              id: 'private-member-list',
              visibility: 'public_limited',
            },
          ]),
        );
      }),
    );

    expect(await draw({ meetup: 'private-event' })).toMatchInlineSnapshot(`
      Object {
        "body": "{\\"error\\":{\\"message\\":\\"Sorry, their members list is private.\\"}}",
        "headers": Object {},
        "statusCode": 404,
      }
    `);
  });

  it('handles unexpected data', async () => {
    jest.spyOn(meetupRandomizer, 'run').mockResolvedValueOnce('unexpected');

    expect(await draw({ meetup: 'some-meetup' })).toMatchInlineSnapshot(`
      Object {
        "body": "{\\"error\\":{\\"message\\":\\"Sorry, we received unexpected data for that request.\\"}}",
        "headers": Object {},
        "statusCode": 404,
      }
    `);
  });

  it('handles a valid Meetup Event', async () => {
    const drawResponse = await draw({ meetup: 'some-meetup' });
    expect(drawResponse).toMatchObject({
      body: expect.any(String),
      headers: expect.any(Object),
      statusCode: 200,
    });

    const body = JSON.parse(drawResponse.body);
    expect(body).toMatchObject({ winners: [expect.any(Object)] });
    expect(body.winners).toEqual(expect.any(Array));
    expect(body.winners.length).toBeGreaterThan(0);
    body.winners.forEach((winner) => {
      expect(winner).toMatchObject({
        name: expect.any(String),
        photoURL: expect.any(String),
        profileURL: expect.any(String),
      });
    });
  });
});
