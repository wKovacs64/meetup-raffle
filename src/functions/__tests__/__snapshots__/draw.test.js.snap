// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw handles Event not found 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"error\\":{\\"message\\":\\"Sorry, I couldn't find any information on that.\\"}}",
      "headers": Object {},
      "statusCode": 404,
    },
  ],
]
`;

exports[`draw handles Event not public 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"error\\":{\\"message\\":\\"Sorry, their members list is private.\\"}}",
      "headers": Object {},
      "statusCode": 404,
    },
  ],
]
`;

exports[`draw handles Meetup not found 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"error\\":{\\"message\\":\\"Sorry, I couldn't find any information on that.\\"}}",
      "headers": Object {},
      "statusCode": 404,
    },
  ],
]
`;

exports[`draw handles a valid Meetup Event 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"winners\\":[{\\"name\\":\\"Tiny Rick\\",\\"photoURL\\":\\"https://i.imgur.com/rgDv1wB.jpg\\",\\"profileURL\\":\\"http://rickandmorty.wikia.com/wiki/Tiny_Rick\\"}]}",
      "headers": Object {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`draw handles invalid requests 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"error\\":{\\"message\\":\\"The \\\\\\"meetup\\\\\\" query parameter is required and must be a string.\\"}}",
      "headers": Object {},
      "statusCode": 400,
    },
  ],
]
`;

exports[`draw handles no upcoming Events found 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"error\\":{\\"message\\":\\"Sorry, I couldn't find any upcoming events.\\"}}",
      "headers": Object {},
      "statusCode": 404,
    },
  ],
]
`;

exports[`draw handles unexpected data 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"error\\":{\\"message\\":\\"Sorry, we received unexpected data for that request.\\"}}",
      "headers": Object {},
      "statusCode": 404,
    },
  ],
]
`;
